exporter/collector/metrics.go

```diff
@@ -272,44 +329,198 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
+					continue
+				}
 			} else {
-				err = me.createTimeSeries(ctx, req)
+				// otherwise export directly
+				err := me.export(ctx, req)
+				if err != nil {
+					errs = append(errs, err)
+				}
 			}
+		}
+	}
+	if len(errs) > 0 {
+		return multierr.Combine(errs...)
+	}
+	return nil
+}
 
-			var st string
-			s := status.Convert(err)
-			st = statusCodeToString(s)
-
-			succeededPoints := len(ts)
-			failedPoints := 0
-			for _, detail := range s.Details() {
-				if summary, ok := detail.(*monitoringpb.CreateTimeSeriesSummary); ok {
-					failedPoints = int(summary.TotalPointCount - summary.SuccessPointCount)
-					succeededPoints = int(summary.SuccessPointCount)
-				}
+func (me *MetricsExporter) export(ctx context.Context, req *monitoringpb.CreateTimeSeriesRequest) error {
+	var err error
+	if me.cfg.MetricConfig.CreateServiceTimeSeries {
+		err = me.createServiceTimeSeries(ctx, req)
+	} else {
+		err = me.createTimeSeries(ctx, req)
+	}
+
+	var st string
+	s := status.Convert(err)
+	st = statusCodeToString(s)
+
+	succeededPoints := len(req.TimeSeries)
+	failedPoints := 0
+	for _, detail := range s.Details() {
+		if summary, ok := detail.(*monitoringpb.CreateTimeSeriesSummary); ok {
+			failedPoints = int(summary.TotalPointCount - summary.SuccessPointCount)
+			succeededPoints = int(summary.SuccessPointCount)
+		}
+	}
+
+	// always record the number of successful points
+	recordPointCountDataPoint(ctx, succeededPoints, "OK")
+	if failedPoints > 0 {
+		recordPointCountDataPoint(ctx, failedPoints, st)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to export time series to GCM: %+v", s)
+	}
+	return err
+}
+
+// readWALAndExport continuously pops CreateTimeSeriesRequest protos from the WAL and
+// tries exporting them. If an export fails due to a network error, it will
+// continually retry the same request until success. Other non-retryable errors
+// drop metrics.
+func (me *MetricsExporter) readWALAndExport(ctx context.Context) error {
+	for {
+		select {
+		case <-me.shutdownC:
+			return nil
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+		bytes, err := me.wal.Read(me.rWALIndex.Load())
+		if err == nil {
+			req := new(monitoringpb.CreateTimeSeriesRequest)
+			if err = proto.Unmarshal(bytes, req); err != nil {
+				return err
 			}
 
-			// always record the number of successful points
-			recordPointCountDataPoint(ctx, succeededPoints, "OK")
-			if failedPoints > 0 {
-				recordPointCountDataPoint(ctx, failedPoints, st)
+			err = me.export(ctx, req)
+			if err != nil {
+				me.obs.log.Warn(fmt.Sprintf("error exporting to GCM: %+v", err))
+			}
+			// retry at same read index if retryable (network) error
+			s := status.Convert(err)
+			if s.Code() == codes.DeadlineExceeded || s.Code() == codes.Unavailable {
+				me.obs.log.Error("non-retryable error, skipping request")
+				continue
 			}
+
+			err = me.wal.TruncateFront(me.rWALIndex.Load())
 			if err != nil {
```

Comment:
Are you testing `err` to see if it's non-retryable here?

Reply:
Retryable errors from GCM are handled in the `continue` statement above (L423). If this line returns an error then it was a problem with the WAL itself